libretroshare/src/services/p3discovery2.cc

```diff
@@ -243,87 +236,80 @@ void p3discovery2::removeFriend(const SSLID &sslId)
 	}
 }
 
-
 PGPID p3discovery2::getPGPId(const SSLID &id)
 {
 	PGPID pgpId;
 	mPeerMgr->getGpgId(id, pgpId);
 	return pgpId;
 }
 
-
 int p3discovery2::tick()
 {
 	return handleIncoming();
 }
 
 int p3discovery2::handleIncoming()
 {
-	RsItem *item = NULL;
+	RsItem* item = nullptr;
 
 	int nhandled = 0;
 	// While messages read
 	while(nullptr != (item = recvItem()))
 	{
-		RsDiscPgpListItem *pgplist = nullptr;
-		RsDiscPgpCertItem *pgpcert = nullptr;
-		RsDiscContactItem *contact = nullptr;
-		RsDiscIdentityListItem *gxsidlst = nullptr;
-		nhandled++;
+		RsDiscPgpListItem* pgplist = nullptr;
+		RsDiscPgpCertItem* pgpcert = nullptr;
+		RsDiscContactItem* contact = nullptr;
+		RsDiscIdentityListItem* gxsidlst = nullptr;
+		RsGossipDiscoveryInviteItem* invite = nullptr;
+		RsGossipDiscoveryInviteRequestItem* inviteReq = nullptr;
+
+		++nhandled;
 
 #ifdef P3DISC_DEBUG
 		std::cerr << "p3discovery2::handleIncoming() Received Message!" << std::endl;
 		item -> print(std::cerr);
 		std::cerr  << std::endl;
 #endif
 
-		if (nullptr != (contact = dynamic_cast<RsDiscContactItem *> (item)))
+		if((contact = dynamic_cast<RsDiscContactItem *>(item)))
 		{
-			if (item->PeerId() == contact->sslId) /* self describing */
+			if (item->PeerId() == contact->sslId)
 				recvOwnContactInfo(item->PeerId(), contact);
-            else
-                processContactInfo(item->PeerId(), contact);
+			else processContactInfo(item->PeerId(), contact);
 		}
-        else  if (nullptr != (gxsidlst = dynamic_cast<RsDiscIdentityListItem *> (item)))
-        {
-            recvIdentityList(item->PeerId(),gxsidlst->ownIdentityList) ;
-            delete item;
-        }
-        else  if (nullptr != (pgpcert = dynamic_cast<RsDiscPgpCertItem *> (item)))
+		else if((gxsidlst = dynamic_cast<RsDiscIdentityListItem *>(item)))
+		{
+			recvIdentityList(item->PeerId(),gxsidlst->ownIdentityList);
+			delete item;
+		}
+		else if((pgpcert = dynamic_cast<RsDiscPgpCertItem *>(item)))
 			recvPGPCertificate(item->PeerId(), pgpcert);
-		else if (nullptr != (pgplist = dynamic_cast<RsDiscPgpListItem *> (item)))
+		else if((pgplist = dynamic_cast<RsDiscPgpListItem *>(item)))
 		{
-			/* two types */
 			if (pgplist->mode == DISC_PGP_LIST_MODE_FRIENDS)
-			{
 				processPGPList(pgplist->PeerId(), pgplist);
-				
-			}
 			else if (pgplist->mode == DISC_PGP_LIST_MODE_GETCERT)
-			{
 				recvPGPCertificateRequest(pgplist->PeerId(), pgplist);
-			}
-			else
-			{
-                		delete item ;
-			}
+			else delete item;
+		}
+		else if((invite = dynamic_cast<RsGossipDiscoveryInviteItem*>(item)))
+			recvInvite(std::unique_ptr<RsGossipDiscoveryInviteItem>(invite));
```

Comment:
so the delete item is implicit in this line? It's explicit everywhere else => hard to read, raises questions, etc. Please don't. It's cool a C++11 feature, but it makes the code harder to read. Furthermore, the way it is written (with explicit delete) shows that the memory ownership is not transferred to the function.

Reply:
No it is not deleted in all places... around retroshare code we have switches where some cases the ownership is transferred to the function and in some cases not and the delete is not there, and one never know if it is a leak or not unless one doesn't follow the whole code...
Instead in this case it is explicit in all the places declaration, implementation and call, that the function take the pointer ownership because it take an unique_ptr and even the compiler enforce that (you cannot pass a raw ptr).
This code is safer and more readable the the old one, I think you just need to get a bit more familiar with C++11 pointer types, they are much safer and don't have performance impact in most of the cases especially unique_ptr have no performance impact at all.

Reply:
In the same file you can see the situation I am talking about, some functions takes ownership and other not, at lines 277, 278, 286, 290, 292 the functions calls seems to take ownership of the pointer, at lines 282, 290 the functions calls seems to not take ownership, and to check if that is correct or not one have to dig deep into the code... while ´recvInvite´ make it explicit that it's taking ownership, and the compiler enforce that too.